compiler/src/dotty/tools/dotc/transform/SpecializeTuples.scala

```diff
@@ -0,0 +1,48 @@
+package dotty.tools
+package dotc
+package transform
+
+import ast.Trees.*, ast.tpd, core.*
+import Contexts.*, Types.*, Decorators.*, Symbols.*, DenotTransformers.*
+import SymDenotations.*, Scopes.*, StdNames.*, NameOps.*, Names.*
+import MegaPhase.MiniPhase
+
+/** Specializes Tuples by replacing tuple construction and selection trees.
+ *
+ * Specifically:
+ * 1. Replaces `(1, 1)` (which is `Tuple2.apply[Int, Int](1, 1)`) and
+ *    `new Tuple2[Int, Int](1, 1)` with `new Tuple2$mcII$sp(1, 1)`.
+ * 2. Replaces `(_: Tuple2[Int, Int])._1` with `(_: Tuple2[Int, Int])._1$mcI$sp`
+ */
+class SpecializeTuples extends MiniPhase:
+  import tpd.*
+
+  override def phaseName: String                 = SpecializeTuples.name
+  override def description: String               = SpecializeTuples.description
+  override def isEnabled(using Context): Boolean = !ctx.settings.scalajs.value
+
+  override def transformApply(tree: Apply)(using Context): Tree = tree match
+    case Apply(TypeApply(fun: NameTree, targs), args)
+        if fun.name == nme.apply && fun.symbol.exists && defn.isSpecializableTuple(fun.symbol.owner.companionClass, targs.map(_.tpe)) =>
+      cpy.Apply(tree)(Select(New(defn.SpecialisedTuple(fun.symbol.owner.companionClass, targs.map(_.tpe)).typeRef), nme.CONSTRUCTOR), args).withType(tree.tpe)
```

Comment:
Shouldn't we test that the `qualifier` of `fun` can safely be elided here?

Reply:
Although contrived, it's possible to construct something like
```scala
val t = { println("foo"); Tuple2 }.apply(1, 2)
```
in which case I think your code will throw away the `println("foo")`. It's possible that this is already broken in `main`, though. I think some stuff with `{ println("foo"); Int }.box(5)` is broken in Scala 2 as well. If that's the case, perhaps at least add a comment?

Reply:
And is there already something that lights up with that prefix?  Is it the "pure" (or is it "idempotent"?) stuff I've noticed but not looked at?